### PR TITLE
Manage host returns from rocBLAS manually

### PR DESCRIPTION
**Pull request by benson31**
Should give us correct, stream-aware host returns in rocBLAS.

Note that if the calling code has already set things to be in device mode, this will do nothing.

### CODE REVIEW

File: src/hydrogen/device/rocBLAS_API.cpp [hunk 1]
```diff
@@ -9,6 +10,99 @@ namespace hydrogen
...
+        {
+            hipStream_t stream;
+            H_CHECK_ROCBLAS(rocblas_get_stream(handle_, &stream));
+#ifdef HYDROGEN_HAVE_CUB
+            H_CHECK_HIP(cub::MemoryPool().DeviceAllocate(
+                            reinterpret_cast<void**>(&device_),
+                            sizeof(T),
+                            stream));
+#else
+            H_CHECK_HIP(hipMallocAsync(
```
Comment (tbennun) [inline src/hydrogen/device/rocBLAS_API.cpp:42]: Probably want to protect that with an ifdef on rocm/hip versions.

File: src/hydrogen/device/rocBLAS_API.cpp [hunk 2]
```diff
@@ -9,6 +10,99 @@ namespace hydrogen
...
+                                           device_,
+                                           sizeof(T),
+                                           hipMemcpyDeviceToHost,
+                                           stream));
+
+                // Clean up device memory
+#ifdef HYDROGEN_HAVE_CUB
+                H_CHECK_HIP(cub::MemoryPool().DeviceFree(device_));
+#else
+                H_CHECK_HIP(hipFreeAsync(device_, stream));
```
Comment (tbennun) [inline src/hydrogen/device/rocBLAS_API.cpp:75]: Same as above

**Review by tbennun — CHANGES_REQUESTED** [review]
Minor change to deal with compatibility

File: src/hydrogen/device/rocBLAS_API.cpp (hunk 1 above)
Comment (benson31) [reply src/hydrogen/device/rocBLAS_API.cpp:42]: The point is valid, but we (implicitly at the moment) require ROCm 5.5.0 or newer, which definitely supports this.

Comment (tbennun) [reply src/hydrogen/device/rocBLAS_API.cpp:42]: Cool, good to know.

**Review by tbennun — APPROVED** [review]

**Review by ndryden — APPROVED** [review]

**Pull request merged**